src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-object.cs

```diff
@@ -0,0 +1,211 @@
+                }
+            }
+            else if (PropertyName is ScriptBlock sb)
+            {
+                string GetScriptBlockResultString(PSObject input)
+                {
+                    var result = sb.DoInvokeReturnAsIs(
+                        useLocalScope: true,
+                        errorHandlingBehavior: ScriptBlock.ErrorHandlingBehavior.WriteToCurrentErrorPipe,
+                        dollarUnder: input,
+                        input: AutomationNull.Value,
+                        scriptThis: AutomationNull.Value,
+                        args: Utils.EmptyArray<object>());
+                    return LanguagePrimitives.ConvertTo<string>(result);
+                }
+
+                AppendValue(GetScriptBlockResultString(_inputObjects[0]));
+
+                for (var index = 1; index < _inputObjects.Count; index++)
+                {
+                    builder.Append(Delimiter);
+                    AppendValue(GetScriptBlockResultString(_inputObjects[index]));
+                }
+            }
+            else
+            {
+                throw new ArgumentException();
+            }
+
+            builder.Append(PostScript);
+            WriteObject(builder.ToString(), false);
+        }
+    }
+
+    internal class JoinItemCompleter : IArgumentCompleter
+    {
+        public IEnumerable<CompletionResult> CompleteArgument(
+            string commandName,
+            string parameterName,
+            string wordToComplete,
+            CommandAst commandAst,
+            IDictionary fakeBoundParameters)
+        {
+            var res = new List<CompletionResult>(10);
+
+            void AddMatching(string completionText, string listText, string toolTip)
+            {
+                if (completionText.StartsWith(wordToComplete, StringComparison.OrdinalIgnoreCase))
+                {
+                    res.Add(new CompletionResult(completionText, listText, CompletionResultType.ParameterValue, toolTip));
+                }
+            }
+
+            AddMatching("', '", "Comma-Space", "', ' - Comma-Space");
+            AddMatching("';'", "Semi-Colon", "';'  - Semi-Colon ");
+            AddMatching("'; '", "Semi-Colon-Space", "'; ' - Semi-Colon-Space");
+            AddMatching($"\"{NewLineText}\"", "Newline", $"{NewLineText} - Newline");
+            AddMatching("','", "Comma", "','  - Comma");
+            AddMatching("'-'", "Dash", "'-'  - Dash");
+            AddMatching("' '", "Space", "' '  - Space");
+            return res;
+        }
+
+        public string NewLineText
+        {
+            get
+            {
+                switch (Environment.NewLine)
+                {
+                    case "\r": return "`r";
+                    case "\n": return "`n";
+                    case "\r\n": return "`r`n";
+                    default: return Environment.NewLine.Replace("\r", "`r").Replace("\n", "`n");
```

Comment:
Seems the switch is superfluous. We could leave only
``'c#
return Environment.NewLine.Replace("\r", "`r").Replace("\n", "`n"); 
```

Reply:
Yes it could. I wrote it the way I did to reduce allocated objects. May not be worth it. 
The string literals will be loaded from metadata and interned. There will only exist one instance of them. The default line will create two new strings each time.

Reply:
Environment.NewLine is static. So I think we should use the usual method
```c#
public static string NewLineText
{
    get
    {
#if UNIX
        return "`n"
#else
        return "`r`n"
#endif
    }
}
```

Reply:
Good point!

Reply:
But I think MacOS uses ``` `r```

Reply:
No, only very old MacOs versions. See my comment below.